workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,421 @@
+# -*- coding: utf-8 -*-
+import logging
+
+logger = logging.getLogger(__name__)
+from random import randint
+from time import sleep
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from typing import Union
+
+
+def prepare_subnodes(
+    subnodes: pd.DataFrame,
+    cities: gpd.GeoDataFrame,
+    regions_onshore: gpd.GeoDataFrame,
+    lau: gpd.GeoDataFrame,
+    head: Union[int, bool] = 40,
+) -> gpd.GeoDataFrame:
+    """
+    Prepare subnodes by filtering district heating systems data for largest systems and assigning the corresponding LAU and onshore region shapes.
+
+    Parameters:
```

Comment:
Could we change the docstring style to numpy? (to be consistent with other PyPSA docstrings)

Reply:
Done